Optimized templated attention to use exp2

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124444
* #124443
* __->__ #124356

0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Chillee committed:
Optimized templated attention to use exp2

[ghstack-poisoned]

drisspg commented on benchmarks/transformer/score_mod.py:
```diff
@@ -198,15 +202,7 @@ def relative_bias(score, b, h, m, n):
 
     def head_bias(score, b, h, m, n):
         return score + 2 * h
-
```
Anti pathogical?

Chillee replied:
yeah, it makes the benchmarks take too long lol.

drisspg commented on test/inductor/test_templated_attention.py:
```diff
@@ -49,18 +49,25 @@ def run_test(self, score_mod: Callable, dtype: torch.dtype = torch.float16):
-        )
+        compiled_error = (golden_out - compiled_out).abs().mean()
+        ref_error = (golden_out - ref_out).abs().mean()
+        if compiled_error > ref_error * 1.1:
```
yup, good change this how we do the rest of the sdpa tests

drisspg replied:
https://github.com/pytorch/pytorch/blob/c9ab9248ce0391c233e7189c0ee71f2aca6a786c/test/test_transformers.py#L93

drisspg commented on torch/_inductor/kernel/templated_attention.py:
```diff
@@ -49,10 +49,8 @@ def sdpa_grid(batch_size, num_heads, num_queries, d_model, meta):
-    # to find the optimal calls for perf/accuracy to tl.dot
     qk_scale = 1.0
-    MATMUL_PRECISION = tl.float16
+    MATMUL_PRECISION = tl.float32
```
this upcasts the bfloat16 /float16 case ?

Chillee replied:
yeah I was trying to get float32 to work, gimme a bit.

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented:
If I enable the new options (ROWS_ASSUME_SAFE and SCORE_MOD_IS_LINEAR), perf on benchmark goes up to 90% of FAV2.

drisspg commented on benchmarks/transformer/score_mod.py:
```diff
@@ -6,25 +6,29 @@
-        globals={"args": args, "kwargs": kwargs, "func": func},
-    )
-    return t0.adaptive_autorange(min_run_time=0.1).median * 1e6
+    return do_bench(lambda: func(*args, **kwargs)) * 1e3
```
This hides kernel launch time overhead but one criticism I have a seen is the increased time to launch, hence why I preferred this.

drisspg commented on torch/_inductor/kernel/templated_attention.py:
```diff
@@ -117,25 +121,26 @@ def sdpa_grid(batch_size, num_heads, num_queries, d_model, meta):
         # -- compute scaling constant ---
         row_max = tl.max(qk, 1)
         m_i_new = tl.maximum(m_i, row_max)
         masked_out_rows = (m_i_new == float("-inf"))
```
Probs anna move this under the assume safe

drisspg commented on benchmarks/transformer/score_mod.py:
```diff
@@ -6,25 +6,29 @@
-    )
-    return t0.adaptive_autorange(min_run_time=0.1).median * 1e6
+    return do_bench(lambda: func(*args, **kwargs)) * 1e3
+    # t0 = benchmark.Timer(
```
maybe remove this

drisspg reviewed (approved):
Some small nits, perf go ↑

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Optimized templated attention to use exp2"


0.705 (vs. FA2) to 0.860 after this change.

<img width="1270" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/d58f57ba-e50e-44ea-8a8a-4f13b8650adf">

to

<img width="1277" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f1945b67-0cfc-463c-a2f6-5812b90677fe">


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

fmassa commented:
cc @danthe3rd , as IIRC this optimization had potentially some implications wrt accuracy

Chillee commented:
@fmassa I guess we'll probably have a variety of configs for controlling it haha.

Chillee commented:
@pytorchbot merge

Chillee commented:
@pytorchbot merge
